gemd/util/impl.py

```diff
@@ -37,41 +39,50 @@ def func(base_obj):
     return
 
 
-def _cached_isinstance_generator(
-        class_or_tuple: Union[Type, Tuple[Type]]) -> Callable[[object], bool]:
+def _cached_isinstance(
+        obj: object,
+        class_or_tuple: Union[Type, Tuple[Union[Type, Tuple[Type]]]]) -> bool:
     """
-    Generate a function that checks and caches an isinstance(obj, class_or_tuple) call.
+    Emulate isinstance builtin to take advantage of functools caching.
 
     Parameters
     ----------
-    class_or_tuple: Union[Type, Tuple[Type]]
-        A single type or a tuple of types
+    obj: object
+
+    class_or_tuple: Union[Type, Tuple[Union[Type, Tuple[Type]]]]
+        A single type, a tuple of types (potentially nested)
 
     Returns
     -------
-    Callable[[object], bool]
-        function with signature function(obj), returning isinstance(obj, class_or_tuple)
+    bool
+        Whether an object is an instance of a class or of a subclass thereof.
 
     """
-    cache = dict()
+    obj_type = type(obj)
+    return _cached_issubclass(obj_type, class_or_tuple)
+
 
-    def func(obj):
-        obj_type = type(obj)
-        if obj_type not in cache:
-            cache[obj_type] = isinstance(obj, class_or_tuple)
-        return cache[obj_type]
+@functools.lru_cache(maxsize=None)
+def _cached_issubclass(
```

Comment:
Is this actually a slow point in the code? What kind of comparisons are the built in `issubclass` and `isinstance` doing that would make caching this so impactful?

Reply:
Just an absurdly high number of them.  This was highlighted as a slow point in previous profiling.  For this PR, I'm just swapping to a less hand-rolled version for legibility.

As an experiment, I encourage you to run the test suite with and without the caching decorator.